### PR TITLE
Fix Afterpay checkout error when shipping information is missing

### LINKED ISSUE #7509: Afterpay checkout fails when shipping information is missing.
### Describe the bug

Afterpay checkout fails when shipping information is missing in the order.

### To Reproduce

1. You'll need to remove the shipping fields from the checkout form. One way to accomplish it is by disabling shipping in `WooCommerce > Settings > General > Shipping location(s)`
2. As a shopper add product(s) for a value greater than 50 USD and proceed to checkout
3. Fill the form, select Afterpay as payment method and place the order

### Actual behavior

The checkout fails and there's an error shown to the shopper:

Screenshot 2023-10-18 at 20 56 05

### Expected behavior

Checkout does not fail with Afterpay

### Additional context

The error seems to originate in this line, when there's missing shipping information in the order.

https://github.com/Automattic/woocommerce-payments/blob/359ea902ca5b5c88952ade7d2310345d620d9854/includes/payment-methods/class-upe-split-payment-gateway.php#L565

We could fallback to the address in the billing details in this case.

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #7509 

#### Changes proposed in this Pull Request

- Validate shipping address from order and fallback to billing address if it is missing or invalid.

#### Testing instructions

* Merchant: Disable shipping in you local store: go to  WooCommerce > Settings > General and set 'Shipping location(s)' as 'Disable shipping& shipping calculations'
* Shopper: add product to cart and proceed to checkout
* Shopper: fill the checkout form and place the order paying with Afterpay
* Observe there are no errors and the checkout proceeds to the Stripe test page.
* Check the shipping data matches the **billing** information provided in previous step.
* Authorize Payment and check that you are redirected back to the order received page.
* Merchant: Enable shipping in your local store (i.e. Ship to all countries)
* Shopper: add product to cart and proceed to checkout
* Shopper: fill the checkout form, select 'Ship to different address' and fill the shipping address details. 
* Shopper: place order.
* Observe there are no errors and the checkout proceeds to the Stripe test page.
* Check the shipping data matches the **shipping** information provided in previous step.
* Authorize Payment and check that you are redirected back to the order received page.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7541` or branch name `fix/7509-afterpay-checkout-fails-when-shipping-information-is-missing` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/7509-afterpay-checkout-fails-when-shipping-information-is-missing) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 8fdf4ae4c53fdf15794c192011c28df86edbcf88
- Build time: 2023-10-26 07:33:08 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -2918,6 +2918,34 @@ public function get_shipping_data_from_order( WC_Order $order ): array {
 		];
 	}
 
+	/**
+	 * Create the shipping data array to send to Stripe when making a purchase, based on order's billing data.
+	 *
+	 * @param WC_Order $order The order that is being paid for.
+	 * @return array          The shipping data to send to Stripe.
+	 */
+	public function get_billing_data_from_order( WC_Order $order ): array {
```
Comment (kalessil) [inline includes/class-wc-payment-gateway-wcpay.php]: Given the logic is not related to the gateway itself, did we consider placing it in `WC_Payments_Order_Service`?

File: includes/class-wc-payments-utils.php [hunk 2]
```diff
@@ -984,4 +984,27 @@ public static function enqueue_style( $handle, $path = '', $deps = [], $version
...
+	 * This function iterates through each required key and checks if that key exists in the data array
+	 * and that its value is not an empty string. If any required key is missing or has an empty string
+	 * value, the function returns false. Otherwise, it returns true.
+	 *
+	 * @param array $data The array containing the data to be validated.
+	 * @param array $required_keys An array of keys that are expected to exist and have non-empty values in the $data.
+	 *
+	 * @return bool Returns true if all required keys are present and their values are not empty strings, false otherwise.
+	 */
+	public static function is_data_valid_against_keys( array $data, array $required_keys ): bool {
```
Comment (kalessil) [inline includes/class-wc-payments-utils.php]: It'd a bit overengineering, let's drop it. I'll share an alternative in another comment.

File: includes/payment-methods/class-upe-split-payment-gateway.php [hunk 3]
```diff
@@ -548,6 +550,33 @@ public function get_stripe_id() {
...
+	 * shipping data for the Afterpay payment request. If neither shipping nor billing data is valid
+	 * for shipping, an exception is thrown.
+	 *
+	 * @param WC_Order                     $order    The order object containing shipping and billing information.
+	 * @param Create_And_Confirm_Intention $request The Afterpay payment request object to set shipping data on.
+	 *
+	 * @throws Invalid_Address_Exception If neither shipping nor billing address is valid for Afterpay payments.
+	 * @return void
+	 */
+	protected function handle_afterpay_shipping_requirement( WC_Order $order, Create_And_Confirm_Intention $request ): void {
```
Comment (kalessil) [inline includes/payment-methods/class-upe-split-payment-gateway.php]: Good refactoring! Do we need to have it protected, or private is possible too?

File: includes/payment-methods/class-afterpay-payment-method.php [hunk 4]
```diff
@@ -15,7 +15,8 @@
  */
 class Afterpay_Payment_Method extends UPE_Payment_Method {
 
-	const PAYMENT_METHOD_STRIPE_ID = 'afterpay_clearpay';
+	const PAYMENT_METHOD_STRIPE_ID         = 'afterpay_clearpay';
+	const SHIPPING_ADDRESS_REQUIRED_FIELDS = [ 'city', 'country', 'line1', 'postal_code', 'state' ];
```
Comment (kalessil) [inline includes/payment-methods/class-afterpay-payment-method.php]: No need for this new constant, will share an alternative in another comment.

File: includes/payment-methods/class-upe-split-payment-gateway.php [hunk 5]
```diff
@@ -548,6 +550,33 @@ public function get_stripe_id() {
...
+		$shipping_data = $this->get_shipping_data_from_order( $order );
+		if ( WC_Payments_Utils::is_data_valid_against_keys( $shipping_data['address'], Afterpay_Payment_Method::SHIPPING_ADDRESS_REQUIRED_FIELDS ) ) {
```
Comment (kalessil) [inline includes/payment-methods/class-upe-split-payment-gateway.php]: There are better approaches than yearly generalization here as we solve extremely narrow task.
We can use something less sophisticated but better contextualized, like:

```php
$check_if_usable = function( array $address ): bool {
	return $address['country'] && $address['state'] && $address['city'] && $address['postal_code'] && $address['line1'];
}

if ( $check_if_usable( $shipping_data['address'] ) ) {
    // ...
}

if ( $check_if_usable( $billing_data['address'] ) ) {
    // ...
}
```

**Review by kalessil — CHANGES_REQUESTED** [review]

File: includes/class-wc-payment-gateway-wcpay.php (hunk 1 above)
Comment (mgascam) [reply includes/class-wc-payment-gateway-wcpay.php]: Good point, I've move this method and also `get_shipping_data_from_order` in this commit https://github.com/Automattic/woocommerce-payments/pull/7541/commits/520dbfae22e6b545e34dd27ee2ae385a2006228d

File: includes/class-wc-payments-utils.php (hunk 2 above)
Comment (mgascam) [reply includes/class-wc-payments-utils.php]: Removed in this commit https://github.com/Automattic/woocommerce-payments/pull/7541/commits/16c534233b8a75a7c727bd6ec870c87f91da4ea4 thanks!

File: includes/payment-methods/class-upe-split-payment-gateway.php (hunk 3 above)
Comment (mgascam) [reply includes/payment-methods/class-upe-split-payment-gateway.php]: Modified to private in this commit https://github.com/Automattic/woocommerce-payments/pull/7541/commits/16c534233b8a75a7c727bd6ec870c87f91da4ea4 thanks!

File: includes/payment-methods/class-afterpay-payment-method.php (hunk 4 above)
Comment (mgascam) [reply includes/payment-methods/class-afterpay-payment-method.php]: Removed in this commit https://github.com/Automattic/woocommerce-payments/pull/7541/commits/16c534233b8a75a7c727bd6ec870c87f91da4ea4 thanks!

File: includes/payment-methods/class-upe-split-payment-gateway.php (hunk 5 above)
Comment (mgascam) [reply includes/payment-methods/class-upe-split-payment-gateway.php]: Indeed, that was a bit too much. Using suggested approach in this commit https://github.com/Automattic/woocommerce-payments/pull/7541/commits/16c534233b8a75a7c727bd6ec870c87f91da4ea4 Thanks for the good pointers (and patience)

**Review by kalessil — APPROVED** [review]
LGTM (code review only)!

File: includes/class-wc-payments-order-service.php [hunk 6]
```diff
@@ -741,6 +741,62 @@ public function attach_intent_info_to_order( $order, $intent_id, $intent_status,
...
+				'postal_code' => $order->get_shipping_postcode(),
+				'city'        => $order->get_shipping_city(),
+				'state'       => $order->get_shipping_state(),
+				'country'     => $order->get_shipping_country(),
+			],
+		];
+	}
+
+	/**
+	 * Create the shipping data array to send to Stripe when making a purchase, based on order's billing data.
```
Comment (dmvrtx) [inline includes/class-wc-payments-order-service.php]:
```suggestion
	 * Create the billing data array to send to Stripe when making a purchase, based on order's billing data.
```

**Review by dmvrtx — APPROVED** [review]
Looks good to me code-wise! Left a slight change in the doc-block.

File: includes/class-wc-payments-order-service.php (hunk 6 above)
Comment (mgascam) [reply includes/class-wc-payments-order-service.php]: Amended here https://github.com/Automattic/woocommerce-payments/pull/7541/commits/05e4c1274b1b9b04301cf093e68a0466382b7e83 thanks!

**Pull request merged**